R/build-sql.R

```diff
@@ -46,3 +46,101 @@ build_sql <- function(..., .env = parent.frame(), con = sql_current_con()) {
   pieces <- purrr::map_chr(enexprs(...), escape_expr, con = con)
   sql(paste0(pieces, collapse = ""))
 }
+
+glue_sql2 <- function(...,
+                      .con,
+                      .sep = "",
+                      .envir = parent.frame(),
+                      .open = "{",
+                      .close = "}",
+                      .na = DBI::SQL("NULL"),
+                      .null = "",
+                      .comment = "#",
+                      .literal = FALSE,
+                      .trim = TRUE) {
+  sql(glue(
+    ...,
+    .sep = .sep,
+    .envir = .envir,
+    .open = .open,
+    .close = .close,
+    .na = .na,
+    .null = .null,
+    .comment = .comment,
+    .literal = .literal,
+    .transformer = sql_quote_transformer(.con),
+    .trim = .trim
+  ))
+}
+
+sql_quote_transformer <- function(connection) {
+  function(text, envir) {
+    collapse_regex <- "[*][[:space:]]*$"
+    should_collapse <- grepl(collapse_regex, text)
+    if (should_collapse) {
+      text <- sub(collapse_regex, "", text)
+    }
+
+    type_regex <- "^\\.(tbl|sql|col|name|from|kw) (.*)"
+    m <- regexec(type_regex, text)
+    is_quoted <- any(m[[1]] != -1)
+    if (is_quoted) {
+      matches <- regmatches(text, regexec(type_regex, text))[[1]]
+
+      type <- matches[[2]]
+      value <- matches[[3]]
+    } else {
+      value <- text
+      type <- "raw"
+    }
+    value <- eval(parse(text = value, keep.source = FALSE), envir)
+    glue_check_collapse(type, should_collapse)
+
+    if (type == "tbl") {
+      if (is_bare_character(value)) {
+        value <- ident(value)
+      }
```

Comment:
I think it's a bit weird to have multiple transformers that do the same thing. Would it be better to just use `name` for all of them?

Do we already have a `as.ident()` helper?

Reply:
In the current implementation this is indeed pretty weird. But the idea for this PR came from my attempt to improve on the handling of table identifiers (see #1205 for some more background). During the implementation of table identifiers build on top of vctrs, I had some issues figuring out where to use the new table identifier and in general what type an argument to e.g. `build_sql()` should be.
This motivated me to be much more explicit about types in `glue_sql2()`. It makes the intent more clear and allows to add some checks on `value` in a future PR.